eth_abi/decoding.py

```diff
@@ -542,24 +541,21 @@ def from_type_str(cls, abi_type, registry):
 
 
 class StringDecoder(ByteStringDecoder):
+    def __init__(self, errors="surrogateescape"):
+        self.bytes_errors = errors
+        super().__init__()
+
     @parse_type_str("string")
     def from_type_str(cls, abi_type, registry):
         return cls()
 
-    @staticmethod
-    def decoder_fn(data):
-        try:
-            value = data.decode("utf-8")
-        except UnicodeDecodeError as e:
-            raise DecodingError(
-                e.encoding,
-                e.object,
-                e.start,
-                e.end,
-                "The returned type for this function is string which is "
-                "expected to be a UTF8 encoded string of text. The returned "
-                "value could not be decoded as valid UTF8. This is indicative "
-                "of a broken application which is using incorrect return types for "
-                "binary data.",
-            ) from e
+    def decode(self, stream):
```

Comment:
To avoid looping in the codec, I think you can optionally pass errors in here:

```suggestion
    def decode(self, stream, errors=None):
        ...
        if errors is None:
            value = self.decoder_fn(data, self.bytes_errors)
        else:
            value = self.decoder_fn(data, errors)
        ...
```
And we may want to add some validation that the error string is in the set of `{'strict', 'surrogateescape', 'backslashreplace', etc.}` but that may be overkill. I assume Python handles it, but it might be nice to add a better error message. Feel free to take or leave it

Reply:
If you register a decoder with "iggnore", for example, the error it gives is `LookupError: unknown error handler name 'iggnore'`, and the stack trace points to a variable of the same name as what they used to register the decoder. It looks clear enough to me, but try registering one with a bad handler name and see if you agree.

Reply:
Sweet, that sounds good to me. Thanks for checking!